### PR TITLE
[SPARK-47398][SQL] Extract a trait for InMemoryTableScanExec to allow for extending functionality

### PR DESCRIPTION
**Pull request by razajafri**
### What changes were proposed in this pull request?
We are proposing to allow the users to have a custom `InMemoryTableScanExec`. To accomplish this we can follow the same convention we followed for `ShuffleExchangeLike` and `BroadcastExchangeLike` 

### Why are the changes needed?
In the PR added by @ulysses-you, we are wrapping `InMemoryTableScanExec` inside `TableCacheQueryStageExec`. This could potentially cause problems, especially in the RAPIDS Accelerator for Apache Spark, where we replace `InMemoryTableScanExec` with a customized version that has optimizations needed by us. This situation could lead to the loss of benefits from [SPARK-42101](https://issues.apache.org/jira/browse/SPARK-42101) or even result in Spark throwing an Exception.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Ran the existing tests 

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by ulysses-you** [comment]
Curious why do you need a customized version of  `InMemoryTableScanExec` ?  Does `CachedBatchSerializer` not enough ? Is it possible to improve the `CachedBatchSerializer` interface if there is a requirement ?

**Comment by razajafri** [comment]
> Curious why do you need a customized version of `InMemoryTableScanExec` ? Does `CachedBatchSerializer` not enough ? Is it possible to improve the `CachedBatchSerializer` interface if there is a requirement ?

Our version of `InMemoryTableScanExec` only supports columnar input/output, if it sees non-columnar data it throws an Exception. We also have our custom metrics being collected in the GPU version. We also have optimized the `convertCachedBatchToColumnarBatch` when `GpuInMemoryTableScanExec` is being used as opposed to the default one.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 1]
```diff
@@ -21,18 +21,37 @@ import org.apache.spark.rdd.RDD
...
 import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.catalyst.plans.logical.Statistics
 import org.apache.spark.sql.catalyst.plans.physical.Partitioning
 import org.apache.spark.sql.columnar.CachedBatch
 import org.apache.spark.sql.execution.{LeafExecNode, SparkPlan, WholeStageCodegenExec}
 import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
 import org.apache.spark.sql.execution.metric.SQLMetrics
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
+trait InMemoryTableScanLike extends LeafExecNode {
```
Comment (tgravescs) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala:32]: please add a description - can just be something basic like what ShuffleExchangeLike has

**Review by tgravescs — COMMENTED** [review]
overall seems like a fine approach, it would be good to get feedback from others more familiar with exact AQE changes to make sure.

**Comment by razajafri** [comment]
@cloud-fan @maryannxue can you please take a look at this?

**Comment by tgravescs** [comment]
+1 lgtm

**Comment by tgravescs** [comment]
merged to master and branch-3.5.

**Comment by dongjoon-hyun** [comment]
Thank you for fixing the regression.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**